src/ufs/mod.rs

```diff
@@ -566,10 +582,32 @@ impl<'a> FileSystem<'a> {
 /// Note that UFS supports "holes"; block-sized and aligned
 /// spans of bytes within a file that are all zeroes are
 /// specially marked, and not backed by allocated blocks.
-#[derive(Clone, Copy, Debug)]
-pub enum Block<'a> {
+enum Block {
     Hole,
-    Sd(&'a [u8]),
+    Sd(io::Sd),
+}
+
+impl Block {
+    fn read(&self, offset: usize, dst: &mut [u8]) {
+        match self {
+            Self::Hole => dst.fill(0),
+            Self::Sd(sd) => {
+                let ptr = sd.as_ptr();
+                let len = sd.len();
+                if offset >= len {
+                    return;
+                }
```

Comment:
i know we talked about reading after seeking past the end of a file being valid _POSIX_, but if we were reading with a bad offset here it's more of a filesystem bug right? `Inode::read` would have had to have calculated a bogus offset into !fixed size! blocks, so something would be severely funky. so i'd think it's worth asserting `offset <= len` rather than a no-op return here. ymmv?

Reply:
Well, less a filesystem bug and more data corruption; consider the loader, for example. It tries to load segments based on data in the ELF program headers: for an ELF executable, for each segment, the corresponding program header has two sets of bounds describing that segment, both denoted by a base/length pair: a memory region to which the segment should be mapped (the base is a memory address, such as a pointer in a virtual address space), and a region in the ELF file that contains the data that goes into each segment (the base of which is an offset into the file). If, for whatever reason, the ELF file is truncated, you could have a short block with an offset that exceeds length.

Reply:
if an ELF had a bad offset, `Inode::read` would return early via `if off > self.size()`. so once we're looking at individual blocks we know the offset _should_ be addressable. that's what makes me think `Block::read` should be more opinionated.

if there's a way we'd be reading from individual blocks without going through `Inode`, definitely don't mind me!